### PR TITLE
[Compression] Add support for deepspeed

### PR DESCRIPTION
**Pull request by Bonytu**
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contrib/compression/utils/evaluator.py [hunk 1]
```diff
@@ -29,6 +29,9 @@
 else:
     TRANSFORMERS_INSTALLED = True
 
+from transformers.trainer_callback import TrainerCallback, TrainerControl, TrainerState
+from transformers.training_args import TrainingArguments
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:33]: put line 32, 33 under line 26

File: nni/contrib/compression/utils/evaluator.py [hunk 2]
```diff
@@ -1057,19 +1093,27 @@ def revert_loss(self):
...
+            def patched_step(_, *args, **kwargs):
+                for task in before_step_tasks:
+                    task()
+                # call origin optimizer step method
+                output = old_step(*args, **kwargs)
+                for task in after_step_tasks:
+                    task()
+                return output
+
+            optimizer.step = types.MethodType(patched_step, optimizer)
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1111]: what does `self.trainer.deepspeed.step` do?

File: nni/contrib/compression/utils/evaluator.py [hunk 3]
```diff
@@ -1057,19 +1093,27 @@ def revert_loss(self):
...
 
-        self.trainer.optimizer.step = types.MethodType(patched_step, self.trainer.optimizer)
+        PatchCallback.on_train_begin = types.MethodType(custom_on_train_begin, PatchCallback)
+
+        # Add Callback into the callback_handler
+        self.trainer.add_callback(PatchCallback)
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1116]: please add a test for `patch_optimizer_step`

File: nni/contrib/compression/utils/evaluator.py [hunk 4]
```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
...
+
+        # will del self._tmp_optimizer and self._tmp_lr_scheduler in `_init_optimizer_helpers`
+        self._tmp_optimizer: Optimizer | Callable[[List[Tensor]], Optimizer] | None = optimizer
+        self._tmp_lr_scheduler: SCHEDULER | Callable[[Optimizer], SCHEDULER] | None = lr_scheduler
+        self._initialization_complete = False
+
+        self.deepspeed_config: DeepSpeedConfig | None = self.process_deepspeed(deepspeed)
+
+    def process_deepspeed(self, config_file_or_dict: str | Dict) -> DeepSpeedConfig:
+        print(f"deepspeed_file: ={config_file_or_dict}")
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1183]: please use logger

File: nni/contrib/compression/utils/evaluator.py [hunk 5]
```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
 
     def get_dummy_input(self) -> Any:
         return self.dummy_input
+
+
+class DeepspeedTorchEvaluator(Evaluator):
+    def __init__(self, training_func: _TRAINING_FUNC, optimizer: Optimizer | Callable[[List[Tensor]], Optimizer] | None,
+                 training_step: _TRAINING_STEP, lr_scheduler: SCHEDULER | Callable[[Optimizer], SCHEDULER] | None,
+                 deepspeed: str | Dict, resume_from_checkpoint_args: Dict | None = None, dummy_input: Any | None = None,
+                 evaluating_func: _EVALUATING_FUNC | None = None):
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/contrib/compression/utils/evaluator.py:1157]: also need a test for this new evaluator

File: nni/contrib/compression/utils/evaluator.py [hunk 6]
```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
 
     def get_dummy_input(self) -> Any:
         return self.dummy_input
+
+
+class DeepspeedTorchEvaluator(Evaluator):
+    def __init__(self, training_func: _TRAINING_FUNC, optimizer: Optimizer | Callable[[List[Tensor]], Optimizer] | None,
+                 training_step: _TRAINING_STEP, lr_scheduler: SCHEDULER | Callable[[Optimizer], SCHEDULER] | None,
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1155]: if `optimizer` and `lr_scheduler` can be wrong, I think we could give them a default `None` value.

File: nni/contrib/compression/utils/evaluator.py [hunk 7]
```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
...
+        # check the validation of optimizer
+        if isinstance(self._tmp_optimizer, Optimizer):
+            assert is_traceable(self._tmp_optimizer)
+        # check the validation of scheduler
+        if isinstance(self._tmp_lr_scheduler, SCHEDULER):
+            assert is_traceable(self._tmp_lr_scheduler)
+
+        # there are 9 cases:
+        # case 1: opt = None, sche = None, depends on the optimizer configuration in deepspeed_config
+        # case 2: opt = Callback, sche = None, ok
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1202]: what is callback mean?

File: nni/contrib/compression/utils/evaluator.py [hunk 8]
```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
...
+
+    def bind_model(self, model: Module, param_names_map: Dict[str, str] | None = None):
+        err_msg = 'Evaluator initialization is not complete, please call `_init_optimizer_helpers` before bind model.'
+        assert self._initialization_complete is True, err_msg
+        assert isinstance(model, Module)
+        if self.model is not None:
+            _logger.warning('Already bound a model, will unbind it before bind a new model.')
+            self.unbind_model()
+
+        self.model = self._rewrap_if_ddp_model(model)
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1254]: seem that this is an assert but not `_rewrap_if_ddp_model`?

File: nni/contrib/compression/utils/evaluator.py [hunk 9]
```diff
@@ -1104,3 +1148,363 @@ def evaluate(self) -> Tuple[float | None, Dict[str, Any]]:
...
+                           'will return None as the model evaluation metric value.'
+                _logger.warning(warn_msg)
+            return nni_used_metric, metric
+        else:
+            return metric
+
+    def get_dummy_input(self) -> Any:
+        return self.dummy_input
+
+    def save_checkpoint(self, save_dir, tag=None, client_state={}, save_latest=True):
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:1451]: Are these save/load functions necessary? If they are, please following the nni docstring format, so that the doc engine could render them.

File: nni/contrib/compression/quantization/lsq_quantizer.py [hunk 10]
```diff
@@ -116,7 +116,16 @@ def register_scale(self):
...
+                # for deepspeed
+                try:
+                    device = next(wrapper.parameters()).device
+                except StopIteration:
+                    try:
+                        device = next(wrapper.buffers()).device
+                    except StopIteration:
+                        # NOTE: this will have risk in model parallel
+                        device = next(self.bound_model.parameters()).device
+                param = torch.nn.Parameter(torch.Tensor([1.0]).to(device))
```
[the remark this answers is not in the document]
Comment (QuanluZhang) [reply nni/contrib/compression/quantization/lsq_quantizer.py:128]: same question

Comment (QuanluZhang) [inline nni/contrib/compression/quantization/lsq_quantizer.py:128]: why deepspeed needs this?

**Comment by QuanluZhang** [comment]
better to briefly share this pr with us in the meeting:)

File: nni/contrib/compression/quantization/lsq_quantizer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Bonytu) [reply nni/contrib/compression/quantization/lsq_quantizer.py:128]: The other quantization algorithms do not have this issue. This problem is mainly caused by Deepspeed's need to allocate buffers for parameters. The previous implementation registered an empty parameter, which would lead to errors.

File: nni/contrib/compression/utils/evaluator.py (hunk 2 above)
Comment (Bonytu) [reply nni/contrib/compression/utils/evaluator.py:1111]: it performs processes such as optimizer.step() and scheduler.step().

File: nni/contrib/compression/quantization/lsq_quantizer.py (hunk 10 above)
Comment (Bonytu) [reply nni/contrib/compression/quantization/lsq_quantizer.py:128]: The other quantization algorithms do not have this issue. This problem is mainly caused by Deepspeed's need to allocate buffers for parameters. The previous implementation registered an empty parameter, which would lead to errors.

File: nni/contrib/compression/quantization/lsq_quantizer.py [hunk 11]
```diff
@@ -116,7 +116,16 @@ def register_scale(self):
...
+                # for deepspeed
+                try:
+                    device = next(wrapper.parameters()).device
+                except StopIteration:
+                    try:
+                        device = next(wrapper.buffers()).device
+                    except StopIteration:
+                        # NOTE: this will have risk in model parallel
+                        device = next(self.bound_model.parameters()).device
+                param = torch.nn.Parameter(torch.Tensor([0.01]).to(device))
```
Comment (QuanluZhang) [inline nni/contrib/compression/quantization/lsq_quantizer.py:128]: does other quantizers also need such changes? or only lsq supports deepspeed evaluator?

File: test/algo/compression/v2_5/utils/bert_ds_config.json [hunk 12]
```diff
@@ -0,0 +1,38 @@
+{
```
Comment (QuanluZhang) [inline test/algo/compression/v2_5/utils/bert_ds_config.json:1]: do we still have folder v2_5?   @J-shang

File: test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
+from pathlib import Path
```
Comment (QuanluZhang) [inline test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py:1]: will this file run in UT or IT?

**Comment by QuanluZhang** [comment]
do we need a doc page to explain the usage of deepspeed?

File: nni/contrib/compression/quantization/lsq_quantizer.py (hunk 11 above)
Comment (Bonytu) [reply nni/contrib/compression/quantization/lsq_quantizer.py:128]: Yes, only the LSQ and LSQ+ quantization algorithms require this because only these two algorithms add extra parameters. DeepSpeed needs to allocate buffers for parameters in advance. If you don't change it according to the above method, you may encounter buffer allocation errors.

Comment (Bonytu) [reply nni/contrib/compression/quantization/lsq_quantizer.py:128]: all quantization and pruning algos support deepspeed evaluator

File: test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py (hunk 13 above)
Comment (Bonytu) [reply test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py:1]: in UT

Comment (QuanluZhang) [reply test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py:1]: which function in this file will be called in UT?

File: test/algo/compression/v2_5/utils/bert_ds_config.json (hunk 12 above)
Comment (J-shang) [reply test/algo/compression/v2_5/utils/bert_ds_config.json:1]: no, this folder will be replaced to `test/algo/compression`

Comment (QuanluZhang) [reply test/algo/compression/v2_5/utils/bert_ds_config.json:1]: great! will be replaced in the following pr, right?

File: test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py (hunk 13 above)
Comment (Bonytu) [reply test/algo/compression/v2_5/utils/test_ds_transfotmers_evaluator.py:1]: I'm sorry, I took a closer look, and this is a unit test (UT). We don't have a separate UT testing folder, so it's placed here.

**Comment by Bonytu** [comment]
> do we need a doc page to explain the usage of deepspeed?

I provided two examples and explained their usage in the evaluator documentation. ）

**Review by QuanluZhang — APPROVED** [review]

File: test/algo/compression/v2_5/utils/bert_ds_config.json (hunk 12 above)
Comment (J-shang) [reply test/algo/compression/v2_5/utils/bert_ds_config.json:1]: yes, in pr #5573

**Review by J-shang — APPROVED** [review]
